// @jcroall/synopsys-sig-node/lib/polaris/input/PolarisInputReader.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var PolarisProxyInfo_1 = __importDefault(require("../model/PolarisProxyInfo"));
var PolarisConnection_1 = __importDefault(require("../model/PolarisConnection"));
var PolarisInputReader = /** @class */ (function () {
    function PolarisInputReader() {
    }
    PolarisInputReader.prototype.getPolarisInputs = function (polaris_url, polaris_token, proxy_url, proxy_username, proxy_password, build_command, should_wait_for_issues, should_changeset, should_changeset_fail) {
        var polaris_proxy_info = undefined;
        if (proxy_url && proxy_url.length > 0 && proxy_username && proxy_username.length > 0 &&
            proxy_password && proxy_password.length > 0) {
            polaris_proxy_info = new PolarisProxyInfo_1.default(proxy_url, proxy_username, proxy_password);
        }
        else {
            polaris_proxy_info = undefined;
        }
        if (polaris_url.endsWith("/") || polaris_url.endsWith("\\")) {
            polaris_url = polaris_url.slice(0, -1);
        }
        return {
            polaris_connection: new PolarisConnection_1.default(polaris_url, polaris_token, polaris_proxy_info),
            build_command: build_command,
            should_wait_for_issues: should_wait_for_issues,
            should_empty_changeset_fail: should_changeset_fail,
            should_populate_changeset: should_changeset
        };
    };
    return PolarisInputReader;
}());
exports.default = PolarisInputReader;
