// @jcroall/synopsys-sig-node/lib/polaris/changeset/ChangeSetEnvironment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var path = require('path');
var ChangeSetEnvironment = /** @class */ (function () {
    function ChangeSetEnvironment(log, env) {
        this.log = log;
        this.env = env;
    }
    ChangeSetEnvironment.prototype.set_enable_incremental = function () {
        this.env["POLARIS_FF_ENABLE_COVERITY_INCREMENTAL"] = "true";
    };
    ChangeSetEnvironment.prototype.is_file_path_present = function () {
        if ("CHANGE_SET_FILE_PATH" in this.env) {
            return true;
        }
        else {
            return false;
        }
    };
    ChangeSetEnvironment.prototype.set_default_file_path = function (cwd) {
        this.env["CHANGE_SET_FILE_PATH"] = path.join(cwd, ".synopsys", "polaris", "changeSetFile.txt");
    };
    ChangeSetEnvironment.prototype.get_file_path = function () {
        return this.env["CHANGE_SET_FILE_PATH"];
    };
    ChangeSetEnvironment.prototype.get_or_create_file_path = function (cwd) {
        if (this.is_file_path_present()) {
            return this.get_file_path();
        }
        else {
            this.set_default_file_path(cwd);
            return this.get_file_path();
        }
    };
    return ChangeSetEnvironment;
}());
exports.default = ChangeSetEnvironment;
